src/server/session.ts

```diff
@@ -2040,4 +2043,50 @@ namespace ts.server {
         response?: {};
         responseRequired?: boolean;
     }
+
+    /* @internal */ // exported for tests only
+    /** Since we will perform the rename in a changed document, line and offset information must be adjusted if the new text adds/deletes newlines. */
+    export function fixupLineAndOffset(oldPosition: protocol.Location, oldText: string, location: number, locationFileName: string, edits: ReadonlyArray<FileTextChanges>): protocol.Location {
+        let { line, offset } = oldPosition;
+        const oldLineStartOfLocation = location - (offset - 1); // offset is 1-based, location is 0-based
+
+        for (const { fileName, textChanges } of edits) {
+            if (fileName !== locationFileName) {
+                continue;
+            }
+
+            // Text change spans are in the *old* document. They apply all at once rather than sequentially.
+            for (const { newText, span: { start, length } } of textChanges) {
+                const newTextNewlines = countNewlines(newText);
+                line += newTextNewlines - countNewlines(oldText.slice(start, start + length));
+                if (start > oldLineStartOfLocation && start < location) {
+                    offset += newText.length - length;
+                    if (newTextNewlines !== 0) {
+                        offset -= newText.lastIndexOf("\n");
+                    }
+                }
+            }
+        }
+
+        return { line, offset };
+    }
+
+    function countNewlines(s: string): number {
+        let count = 0;
+        for (let i = 0; i < s.length; i++) {
+            switch (s.charCodeAt(i)) {
+                case CharacterCodes.carriageReturn:
+                    count++;
+                    i++;
+                    // Skip a "\n" after a "\r" because "\r\n" is only one newline. But "\r\r" is two.
+                    if (s.charCodeAt(i) === CharacterCodes.carriageReturn) {

```

Comment:
Do you need a bounds check in case the last character is `\r`?

Reply:
Should be fixed now